### ISSUE TITLE
Unserialize Drawings 

### ISSUE THREAD

**Issue by workingbuddy10** [amcharts/amcharts5#1297]
When I use Delete ALL Drawings button, it deletes the unserializeDrawings  drawings too!

I only to delete the unserializeDrawings  by method you told me of _disposeIndex_ .

I want : 
 - User made drawings should be deleted by the Earaser & Delete All drawings button ONLY.
 - And unserializeDrawings should be deleted by disposeIndex method which you introduced ONLY.
 
 
 
 
And the second issue I am facing in a case is : 

When I draw a drawing (Line) on the unserializeDrawing ( which is also a line) then dispose index is not deleting the drawing.

The Line which is marked in Yellow is unserializeDrawing and the Red marked line is user drawn.
![image](https://github.com/amcharts/amcharts5/assets/137867461/fc3dfe33-544a-40fb-9d3d-d5d8def6187b)

I am deleting like this :
```
     deleteAutoIndicator(deleteInd) {
        
        this.chartInstance.mainPanel.drawings.each(function (series) {
            if (series.className == "SimpleLineSeries") {
                deleteInd = series;
            }
        })
        deleteInd.disposeIndex(2);

    }
```

Full conversation #1280

**Comment by workingbuddy10** [comment]
And If the below is the data of the drawings of the tool and I want to Only delete the 1st "Horizontal Ray" tool only. How can I do it?

```
    {
        "__tool": "Line",
        "__panelIndex": 0,
        "__drawing": [
            {
                "stroke": {
                    "type": "Template",
                    "settings": {
                        "stroke": {
                            "type": "Color",
                            "value": "#882dff"
                        },
                        "strokeOpacity": 1,
                        "strokeDasharray": [],
                        "strokeWidth": 2
                    }
                },
                "fill": {
                    "type": "Template",
                    "settings": {
                        "fill": {
                            "type": "Color",
                            "value": "#ad6eff"
                        },
                        "fillOpacity": 0.5
                    }
                },
                "index": 1,
                "corner": "e",
                "__parse": true
            },
            {
                "valueY": -94.2,
                "valueX": 1700571600000,
                "index": 1,
                "corner": "p1",
                "__parse": true
            },
            {
                "valueY": -94.2,
                "valueX": 1700665200000,
                "index": 1,
                "corner": "p2",
                "__parse": true
            }
        ]
    },
    {
        "__tool": "Line",
        "__panelIndex": 0,
        "__drawing": [
            {
                "stroke": {
                    "type": "Template",
                    "settings": {
                        "stroke": {
                            "type": "Color",
                            "value": "#882dff"
                        },
                        "strokeOpacity": 1,
                        "strokeDasharray": [],
                        "strokeWidth": 2
                    }
                },
                "fill": {
                    "type": "Template",
                    "settings": {
                        "fill": {
                            "type": "Color",
                            "value": "#ad6eff"
                        },
                        "fillOpacity": 0.5
                    }
                },
                "index": 1,
                "corner": "e",
                "__parse": true
            },
            {
                "valueY": -97.5,
                "valueX": 1700578800000,
                "index": 1,
                "corner": "p1",
                "__parse": true
            },
            {
                "valueY": -97.5,
                "valueX": 1700661600000,
                "index": 1,
                "corner": "p2",
                "__parse": true
            }
        ]
    },
    {
        "__tool": "Horizontal Line",
        "__panelIndex": 0,
        "__drawing": [
            {
                "stroke": {
                    "type": "Template",
                    "settings": {
                        "stroke": {
                            "type": "Color",
                            "value": "#882dff"
                        },
                        "strokeOpacity": 1,
                        "strokeDasharray": [],
                        "strokeWidth": 2
                    }
                },
                "fill": {
                    "type": "Template",
                    "settings": {
                        "fill": {
                            "type": "Color",
                            "value": "#ad6eff"
                        },
                        "fillOpacity": 0.5
                    }
                },
                "index": 1,
                "corner": "e",
                "__parse": true
            },
            {
                "valueY": -102.6,
                "valueX": 1697709600000,
                "index": 1,
                "corner": "p1",
                
                "__parse": true
            },
            {
                "valueY": -102.6,
                "valueX": 1705701600000,
                "index": 1,
                "corner": "p2",
                "__parse": true
            }
        ]
    },
    {
        "__tool": "Horizontal Line",
        "__panelIndex": 0,
        "__drawing": [
            {
                "stroke": {
                    "type": "Template",
                    "settings": {
                        "stroke": {
                            "type": "Color",
                            "value": "#882dff"
                        },
                        "strokeOpacity": 1,
                        "strokeDasharray": [],
                        "strokeWidth": 2
                    }
                },
                "fill": {
                    "type": "Template",
                    "settings": {
                        "fill": {
                            "type": "Color",
                            "value": "#ad6eff"
                        },
                        "fillOpacity": 0.5
                    }
                },
                "index": 1,
                "showExtension": true,
                "corner": "e",
                "__parse": true
            },
            {
                "valueY": -99.7,
                "valueX": 1697709600000,
                "corner": "p1",
                "index": 1,
                "__parse": true
            },
            {
                "valueY": -99.7,
                "valueX": 1705701600000,
                "corner": "p2",
                "index": 1,
                "__parse": true
            }
        ]
    },
```

**Comment by martynasma** [comment]
There's a bug which causes drawing series to duplicate for each activation of the tool, which is part of the problem. We'll fix it in the next version.

The second part of the problem is the serialized data that you have.

The tool is duplicated twice (I'm assuming it's the result of the aforementioned bug), but there should be only one entry per tool/panel, e.g.:

```JavaScript
[{
  "__tool": "Line",
  "__panelIndex": 0,
  "__drawing": [{
    "stroke": {
      "type": "Template",
      "settings": {
        "stroke": {
          "type": "Color",
          "value": "#882dff"
        },
        "strokeOpacity": 1,
        "strokeDasharray": [],
        "strokeWidth": 2
      }
    },
    "fill": {
      "type": "Template",
      "settings": {
        "fill": {
          "type": "Color",
          "value": "#ad6eff"
        },
        "fillOpacity": 0.2
      }
    },
    "index": 1,
    "showExtension": true,
    "corner": "e",
    "__parse": true
  }, {
    "valueY": 583.9,
    "valueX": 1623816973404,
    "corner": "p1",
    "index": 1,
    "__parse": true
  }, {
    "valueY": 579.3,
    "valueX": 1633515242141,
    "corner": "p2",
    "index": 1,
    "__parse": true
  }, {
    "stroke": {
      "type": "Template",
      "settings": {
        "stroke": {
          "type": "Color",
          "value": "#882dff"
        },
        "strokeOpacity": 1,
        "strokeDasharray": [],
        "strokeWidth": 2
      }
    },
    "fill": {
      "type": "Template",
      "settings": {
        "fill": {
          "type": "Color",
          "value": "#ad6eff"
        },
        "fillOpacity": 0.2
      }
    },
    "index": 2,
    "showExtension": true,
    "corner": "e",
    "__parse": true
  }, {
    "valueY": 431.2,
    "valueX": 1626777782575,
    "corner": "p1",
    "index": 2,
    "__parse": true
  }, {
    "valueY": 463.7,
    "valueX": 1636069225698,
    "corner": "p2",
    "index": 2,
    "__parse": true
  }]
}]
```

Pay attention to the `index` setting in the data.

`index` is a bit misleading. It should be treated like a unique ID of the element, rather than its place representation in the list.

In any case, I suggest you wait for the next release before re-generating your data.

It will also have `drawingSeries` property exposed on `DrawingControl` which will make it easier to locate series for the specific tool (https://github.com/amcharts/amcharts5/issues/1280#issuecomment-1903724649).

**Comment by workingbuddy10** [comment]
Just wanted to know behavior of  index.
- How it is defined or generated?
- Is it any random number given to drawing ? Or if I say I have drawn 3 Horizontal lines then the first line index will be 0, then 1,2 for the other lines?

@martynasma

**Comment by workingbuddy10** [comment]
Let's say I have serialized  Horizontal line with index 0.

And user draw one Horizontal again. 
- What will be the index of this user made drawing.
- And If I try to dispose the Line with index, then which line will delete? User made drawing or serialized drawing.

**Comment by martynasma** [comment]
> How it is defined or generated?

The first line gets assigned an index of 1. The next one - 2.

Deleting 1 will not update the second line's index. It will remain index: 2.

Please note that index (id) also carries over when serializing/unserializing.

If you unserialize 2 lines, they will have the same index as they were exported.

When user adds new line, it will take the next number, e.g. 3.

**Comment by workingbuddy10** [comment]
So for example I have a json file which contains the template/data for un-serializing Drawings of product  X.
And the json file contains 2 horizontal lines which have the index as 1,2.

The user came and draw a horizontal line, and after that he unserialized the drawings ( from the json file data).

- So the user drawn horizontal line index will be 1 ?
- And he  un-serialize 2 horizontal lines index as 1,2?

Then I tried to dispose the index of horizontal line whose index is 1, which line will be deleted? 
The user-made or the  un-serialized one.


Note : I didn't directly  un-serialize the drawings from product X, I know what template the unserializeDrawings expects on the basis of that I have already generated that json file.

@martynasma

**Comment by martynasma** [comment]
> So the user drawn horizontal line index will be 1 ?

No, it will be 3.

**Comment by workingbuddy10** [comment]
@martynasma One question, I want to ask regarding index of drawings that can we give the "alpha numeric"  values to the index?

If not, then in the next version can you add this? It would solve some of problems like: 
- In the case of deleting un-serialize drawings, I can dispose index of "alpha numeric"  values of the drawings index. 


  For example I made a head & shoulder pattern with different lines and  un-serialize it which have index as _"headnShoulder 1"_, later when user want to remove Only head & shoulder pattern drawings then I can pass dispose index _"headnShoulder 1"_. By this no other user drawings will be disturbed.

**Comment by martynasma** [comment]
> One question, I want to ask regarding index of drawings that can we give the "alpha numeric" values to the index?

No, it can only be a number.

At this time we do not plan on changing this.

**Comment by workingbuddy10** [comment]
> No, it will be 3.

@martynasma  What will happen in the case of Data Save Control?

For example I have an file which contains un-serialize drawings of product X.

Now user have 3 options related to drawings :
- Can draw using drawing tools.
- Can retrieve drawings from local storage which he had saved once.
- Can choose the option to view the drawings of the file ( which contains un-serialize drawing set of data).

Let's see a sample real-life case: 

X product un-serialize File content:  It will have Template of horizontal ray which has 3 horizontal ray's with the indexes 0,1,2. This file I have generated from a script as I already know what template unserializeDrawings accepts.

1.  Now user came and draw  a horizontal ray ( which will have an index as 0), and then he saved it in local storage using DataSaveControl feature. 
2. Now he removed the ray from the chart ( Note: it is already saved in local storage).
3. Again he used the drawing tools to draw a horizontal ray.  Let's Assume the index is 1 of the ray.
   Question 1:  what will be the index of the ray which user made at step 3?
4.  Now he Retrieved the horizontal ray which is stored in local storage via DataSaveControl feature.
5. And Lastly he also unserializeDrawings of the file too which contains  3 horizontal ray's with the indexes 0,1,2.


Question 2: 
Now you can see there are total 5 Horizontal ray's on the mainpanel/chart. 
- One the user which he had made. (with index 1). _(step 3)_
- One which he retrieved from local storage. ( with index 0).  _(step 1)_
- And 3 other's from unserializeDrawings. ( with index 0,1,2). _( step 5)_

You can see there are horizontal ray's which has same index.

How the index of drawings will behave and change. As I can Only pass index to delete the drawings ( disposeIndex method)?
As passing index 0 might delete multiple horizontal ray, due to common index.

**Comment by martynasma** [comment]
You're right. Unserializing drawings when there already are user-drawn drawings of the same type might result in duplicate indexes.

We'll need to address this. Would you mind posting it as a separate issue so that we can track it better?

**Comment by martynasma** [comment]
Fixed in [5.8.0](https://github.com/amcharts/amcharts5/releases/tag/5.8.0).

## [5.8.0] - 2024-02-01

### Added
- New container type `ZoomableContainer` added, which allows adding zoom capabilities to virtually any chart. [More info](https://www.amcharts.com/docs/v5/concepts/common-elements/containers/#Zoomable_container).
- New class `ZoomTools` added. Can be used to quickly add zoom support for elements compatible with `IZoomable` interface, e.g. `ZoomableContainer`.
- New read-only property of `DrawingControl`: `drawingSeries`. Contains an object where key is a drawing tool name and values are array with references to actual drawing series.

### Changed
- Default value for `strokeOpacity` for the grid line of a `StockChart` changed from `0` to `0.4` so that it would differ from regular grid lines.

### Fixed
- Resizing stock chart's panels after moving them up/down was not working properly.
- `Hierarchy.addChildData` was duplicating node's children which were added before.
- If `topDepth` was set to `1` on `Partition`, initially the chart was showing not all levels of nodes.
- When clicking on particular partitions of a `Partion` chart, zoom animation was not playing.
- `DateRangeSelector` was showing extra day in automatically-calculated end date.
- `DateRangeSelector` was zooming to the start of the selected day, instead of end.
- Tooltip background color passed to `HeatLegend.showValue(value, text, color)` was being ignored.
- `ClusteredPointSeries` was not showing charts if they were added to clustered bullets.
- Fixed issue with `DrawingControl` which was unnecessarily duplicating drawing series with each tool selection / API call.
- Using "1M" button in `PeriodSelector` was not always selecting full month.
- Restoring drawings with `DrawingControl.unserializeDrawings()` will now correctly update indexes of the restored drawings based on current drawing counter.
- In some cases `StockChart` was taking a data item which was just outside of zoom range when calculating aggregate values for its series, affecting percent change values and some other aggregates.
- Legend value was not being updated when series axis' range changed. This caused `legendRangeLabelText` to not being updated (if set) when chart was being zoomed-in/out.
- If a type of `StockChart` series was changed with data grouping enabled, some duplicate candlesticks or ohlc sticks were displayed until first zoom.
- In some rare cases chart would show incorrect last segment of a `LineSeries` when in percent mode.

[Full change log](https://github.com/amcharts/amcharts5/blob/master/packages/shared/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
